Extensions/Physics3DBehavior/Physics3DRuntimeBehavior.ts

```diff
@@ -0,0 +1,1873 @@
+    'Physics3D::Physics3DBehavior',
+    gdjs.Physics3DRuntimeBehavior
+  );
+
+  export namespace Physics3DRuntimeBehavior {
+    /**
+     * Allow extensions relying on the 3D physics to customize its
+     * behavior a bit.
+     */
+    export interface Physics3DHook {
+      /**
+       * Called before the physics engine step.
+       */
+      doBeforePhysicsStep(timeDelta: float): void;
+    }
+
+    export interface BodyUpdater {
+      createAndAddBody(): Jolt.Body;
+      updateObjectFromBody(): void;
+      updateBodyFromObject(): void;
+      recreateShape(): void;
+    }
+
+    export class DefaultBodyUpdater {
+      behavior: gdjs.Physics3DRuntimeBehavior;
+
+      constructor(behavior: gdjs.Physics3DRuntimeBehavior) {
+        this.behavior = behavior;
+      }
+
+      createAndAddBody(): Jolt.Body {
+        const { behavior } = this;
+        const { _sharedData } = behavior;
+
+        const shape = behavior.createShape();
+        const bodyCreationSettings = new Jolt.BodyCreationSettings(
+          shape,
+          behavior.getPhysicsPosition(_sharedData.getRVec3(0, 0, 0)),
+          behavior.getPhysicsRotation(_sharedData.getQuat(0, 0, 0, 1)),
+          behavior.bodyType === 'Static'
+            ? Jolt.EMotionType_Static
+            : behavior.bodyType === 'Kinematic'
+            ? Jolt.EMotionType_Kinematic
+            : Jolt.EMotionType_Dynamic,
+          behavior.getBodyLayer()
+        );
+        bodyCreationSettings.mMotionQuality = behavior.bullet
+          ? Jolt.EMotionQuality_LinearCast
+          : Jolt.EMotionQuality_Discrete;
+        bodyCreationSettings.mAllowedDOFs = behavior.fixedRotation
+          ? Jolt.EAllowedDOFs_TranslationX |
+            Jolt.EAllowedDOFs_TranslationY |
+            Jolt.EAllowedDOFs_TranslationZ
+          : Jolt.EAllowedDOFs_All;
+        bodyCreationSettings.mFriction = behavior.friction;
+        bodyCreationSettings.mRestitution = behavior.restitution;
+        bodyCreationSettings.mLinearDamping = behavior.linearDamping;
+        bodyCreationSettings.mAngularDamping = behavior.angularDamping;
+        bodyCreationSettings.mGravityFactor = behavior.gravityScale;
+
+        const bodyInterface = _sharedData.bodyInterface;
+        const body = bodyInterface.CreateBody(bodyCreationSettings);
+        Jolt.destroy(bodyCreationSettings);
+
+        bodyInterface.AddBody(body.GetID(), Jolt.EActivation_Activate);
+        return body;
+      }
+
+      updateObjectFromBody() {
+        const { behavior } = this;
+        const { owner3D, _body } = behavior;
+        // Copy transform from body to the GD object.
+        // It's possible the behavior was either deactivated or the object deleted
+        // just before this doStepPreEvents (for example, another behavior deleted
+        // the object during its own doStepPreEvents). If the body is null, we just
+        // don't do anything (but still run the physics simulation - this is independent).
```

Comment:
The parenthesis comment is useless now. And in 2D only, since you've reworked how this code works